docs/proposals/rate-limit.md

```diff
@@ -0,0 +1,45 @@
+# Rate Limit
+
+## Context
+
+Today the only way to support Rate Limiting using Kuma – create a ProxyTemplate. Current proposal introduces a new 
+policy Rate Limit.
+
+## Requirements
+
+Provide a way to configure local rate limit both for [TCP](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto#envoy-v3-api-msg-extensions-filters-network-local-ratelimit-v3-localratelimit) and [HTTP](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto#envoy-v3-api-msg-extensions-filters-http-local-ratelimit-v3-localratelimit).
+
+## Configuration
+
+```yaml
+type: RateLimit
+name: rt-1
+mesh: default
+sources:
+- match:
+    kuma.io/service: web
+destinations:
+- match:
+    kuma.io/service: backend
+conf:
+  tcp:
+    connections: 100 # how many connections are allowed per 'period'
+    period: 1s
+    max: 1000 # if 'max' is not specified we can take it 10 times more than 'connections'
+  http:
+    requests: 100 # how many requests are allowed per 'period'
+    period: 1s
+    max: 1000 # if 'max' is not specified we can take it 10 times more than 'requests'
+    status: 401
+    headers:
+     - key: x-local-rate-limit
+       value: 'true'
+       append: false
+```
+
+## Implementation details
+
+Since it is a local rate limiting we have to set it on the destination side. It could be challenging to support 
+`source` selector, but still quite possible for HTTP. We already set header `x-kuma-tags` to identify the source for
+FaultInjections, I believe the same approach will work for RateLimit. For TCP most likely we’ll have a limitation 
+on `source` selector unless it’s possible to somehow use SANs but still this is mTLS only. 
```

Comment:
@subnetmarco 

What is the main use case that users and customers are looking for?

Is it about throttling `web` so it won't send more than 100 req/s to `backend` if app is misbehaving or is it about protecting `backend` that it does not receive more than X req/s?

If former, then we can go with Ilya's proposal, but if it is about protecting `backend` then it should be
```
type: RateLimit
name: rt-1
mesh: default
selectors:
- match:
    kuma.io/service: backend # apply on backend for any traffic
conf:
  tcp:
    connections: 100 # how many connections are allowed per 'period'
    period: 1s
  http: ...
```

To give more examples with the current approach

1.
```
---
type: RateLimit
name: rt-1
mesh: default
sources:
- match:
    kuma.io/service: web
destinations:
- match:
    kuma.io/service: backend
conf:
  http:
    requests: 100 # how many requests are allowed per 'period'
    period: 1s
---
type: RateLimit
name: rt-1
mesh: default
sources:
- match:
    kuma.io/service: another-client
destinations:
- match:
    kuma.io/service: backend
conf:
  http:
    requests: 100 # how many requests are allowed per 'period'
    period: 1s
```
web AND another-client are able to execute 100 req/s.

2.
```
type: RateLimit
name: rt-1
mesh: default
sources:
- match:
    kuma.io/service: '*'
destinations:
- match:
    kuma.io/service: backend
conf:
  http:
    requests: 100 # how many requests are allowed per 'period'
    period: 1s
```
any instance of a service can execute 100 req/s

Reply:
but I proposed to apply rate-limiting on the destination side, selecting by `sources` might be tricky but I think it's possible to implement at least for HTTP

Reply:
@jakubdyszkiewicz @subnetmarco I reflected this discussion as 3 different options in the proposal

Reply:
Rate limiting is for protection of the destination. Although the destination service may want to assign different limits to different sources. Use cases are:

* `backend` will only handle 100 req/s from any source.
* `backend` will only handle 100 req/s from any source, but a specific `backend-app` source should be able to make 200req/s.

Reply:
Applying the rate limiter on the destination only will render it not usable with External Services. I'd definitely consider applying the rate limiter on the source, but use the Global Rate limiter instead of the Local one -> https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/global_rate_limiting
This way we have a full view of the number of RPS within the full system instead of just a single point. Consider this case:
 - Sevice A is rate limited to 100RPS
 - Service A has 3 instances A1, A2, A3
 - applying a local rate limiter on the destination will result in 300 RPS overall
 - applying a global rate limiter on the source will result in 100 RPS always, no matter the number of clients or number of instances of A